### ISSUE TITLE
Discussion on new CI

### ISSUE THREAD

**Issue by yarnabrina** [sktime/sktime#5706]
# Context

The existing (old) CI in `sktime` mainly use `all_extras_pandas2` dependency, except for few specific dedicated jobs. This often leads to the installation of non-latest package versions, so documented upper dependencies is actually not tested in CI's triggered by the pull requests.

There is also the case that all unit tests are triggered for any type of changes, for example a trivial change in a docstring or `README.md` or etc. still triggers full CI. @fkiraly introduced two optimisations for these scenarios:

1. **sub-sampling**: instead of all tests running on all combinations of operating systems and python versions, only a randomly selected subset of tests run in each environment, with a guarantee that each test runs at least once each operating system and at least once on each python version
2. **differential testing**: instead of running all tests, only tests corresponding to estimators present in the modified files get executed

These optimisations significantly reduce the time, but the CI still triggers the tests first for all types of changes and uses (possibly significantly) older versions of the soft-dependencies. It also reduces the code coverage because some specific combinations are being intentionally skipped with objective to minimise CI duration. Another issue observed was that there was no way to verify whether different simultaneous PR's are compatible with one another in the sense they do not introduce any breaking by interaction, even though independently both are non-breaking.

To minimise (and hopefully remove, but not yet) these problems, new CI was proposed in #5101 and implemented in few recent PR's (#5136, #5304, #5375, #5427, #5437, #5498, #5638).

# Implementation

Some recent discussions in Discors seems to suggest that is possibly some misunderstanding in the implementation details, and what is happening in which scenario. The following is a high level outline for logics/introduced in new CI.

## Actions

1. `.github/actions/test-base/action.yml`
    - **current name**: `test base sktime`
    - **current description**: `run unit tests on base interface of sktime framework`
    - **objective**
        - install `sktime` in the specified python version along with testing dependencies
        - run all unit tests in that specified component with specified choices for sub-sampling estimators and differential testing
2. `.github/actions/test-component/action.yml`
    - **current name**: `test specific sktime component`
    - **current description**: `run unit tests on individual component of sktime framework`
    - **objective**
        - install `sktime` with the optional dependency for the specified `sktime` component in the specified python version along with testing dependencies
        - run all unit tests in that specified component with specified choices for sub-sampling estimators and differential testing
3. `.github/actions/validate-extra/action.yml`
    - **current name**: `test specific sktime extra`
    - **current description**: `install individual extra of sktime`
    - **objective**
        - install `sktime` with the specified optional dependency in the specified python version
        - goal is to check if installation succeeds or not

## Workflows

1. `.github/workflows/test_all.yml`
    - **current name**: `test everything`
    - **jobs**
        - `code_quality`
            - **current name**: `validate code quality`
            - **objective**
                - install `pre-commit`
                - run all configured `pre-commit` hooks on all files tracked by `git`
                - identify sub-directories of `sktime` that lack a `__init__.py` file
        - `test_base`
            - **current name**: `test base component`
            - **objective**
                - stop if `code_quality` has failed
                - for all combinations in the Cartesian product of supported operating systems and supported python versions
                    - install `sktime` along with testing dependencies
                    - run all unit tests in `base` component of `sktime` with both sub-sampling and differential testing disabled
                    - upload code coverage information with relevant flags
        - `test_components_with_extras`
            - **current name**: `test components with extras`
            - **objective**
                - stop if `code_quality` has failed
                - for all combinations in the Cartesian product of supported operating systems, supported python versions and all `sktime` components with dedicated optional dependencies
                    - install `sktime` with dedicated optional dependency for the selected component along with testing dependencies
                    - run all unit tests in selected component of `sktime` with both sub-sampling and differential testing disabled
                    - upload code coverage information with relevant flags
        - `test_components_without_extras`
            - **current name**: `test components without extras`
            - **objective**
                - stop if `code_quality` has failed
                - for all combinations in the Cartesian product of supported operating systems and supported python versions
                    - install `sktime` with `all_extras_pandas2` optional dependency along with testing dependencies
                    - run all unit tests in `sktime` components that lack dedicated optional dependencies (other than `base` and `datasets`) with both sub-sampling and differential testing disabled
                    - upload code coverage information with relevant flags
2. `.github/workflows/test_base.yml`
    - **current name**: `test base component`
    - **jobs**
        - `detect`
            - **current name**: `detect changes in base component`
            - **objective**
                - identify if there is any change in any of the `base` component of `sktime`
        - `test`
            - **current name**: `test base component`
            - **objective**
                - stop if no change is detected
                - otherwise for all combinations in the Cartesian product of supported operating systems and supported python versions
                    - install `sktime` along with testing dependencies
                    - run all unit tests in `base` component of `sktime` with both sub-sampling and differential testing enabled
                    - upload code coverage information with relevant flags
3. `.github/workflows/test_code_quality.yml`
    - **current name**: `validate code quality workflow`
    - **jobs**
        - `code_quality`
            - **current name**: `test code quality`
            - **objective**
                - install `pre-commit`
                - identify which files are changed
                - run all configured `pre-commit` hooks only on modified files
                - identify sub-directories of `sktime` that lack a `__init__.py` file
        - `detect`
            - **current name**: `detect changes`
            - **objective**
                - identify if there is any change in either `base` or `datasets` component
                - identify if there is any change in `pyproject.toml`
        - `test_install_extras`
            - **current name**: `validate extra installs`
            - **objective**
                - stop if no change is detected in `pyproject.toml`
                - otherwise trigger `.github/workflows/validate_extras.yml` workflow
        - `unit_test_base`
            - **current name**: `run unit tests for base framework`
            - **objective**
                - stop if `code_quality` has failed
                - otherwise trigger `.github/workflows/test_base.yml` workflow
        - `unit_test_components_with_extras`
            - **current name**: `run unit tests for individual components with extras`
            - **objective**
                - stop if `code_quality` has failed
                - otherwise trigger `.github/workflows/test_components_with_extras.yml` workflow
        - `unit_test_components_without_extras`
            - **current name**: `run unit tests for individual components without extras`
            - **objective**
                - stop if `code_quality` has failed
                - otherwise trigger `.github/workflows/test_components_without_extras.yml` workflow
        - `unit_test_datasets`
            - **current name**: `run unit tests for datasets`
            - **objective**
                - stop if no change is detected in both `datasets` and `base` componens of `sktime`
                - otherwise trigger `.github/workflows/validate_datasets.yml` workflow
4. `.github/workflows/test_components_with_extras.yml`
    - **current name**: `test components with extras`
    - **jobs**
        - `detect`
            - **current name**: `detect changes in components with extras`
            - **objective**
                - for all `sktime` components with dedicated optional dependencies
                    - identify if there is any change in the selected `sktime` component or in the `base` component
        - `test`
            - **current name**: `test components with extras`
            - **objective**
                - for all `sktime` components with dedicated optional dependencies
                    - stop if no change is detected in both the selected component and `base` component
                    - otherwise for all combinations in the Cartesian product of supported operating systems and supported python versions
                        - install `sktime` with dedicated optional dependency for the selected component along with testing dependencies
                        - run all unit tests in that selected `sktime` component with both sub-sampling and differential testing enabled
                        - upload code coverage information with relevant flags
5. `.github/workflows/test_components_without_extras.yml`
    - **current name**: `test components without extras`
    - **jobs**
        - `detect`
            - **current name**: `detect changes in components without extras`
            - **objective**
                - identify if there is any change in any of the `sktime` components that lack a dedicated optional dependency (other than `datasets`)
        - `test_components_without_extras`
            - **current name**: `test components without extras`
            - **objective**
                - stop if no change is detected
                - otherwise for all combinations in the Cartesian product of supported operating systems and supported python versions
                    - install `sktime` with `all_extras_pandas2` optional dependency along with testing dependencies
                    - run all unit tests in `sktime` components that lack dedicated optional dependencies (other than `base` and `datasets`) with both sub-sampling and differential testing enabled
                    - upload code coverage information with relevant flags
6. `.github/workflows/validate_datasets.yml`
    - **current name**: `datasets validation`
    - **jobs**
        - `test_internal_data`
            - **current name**: `test data shipped with sktime`
            - **objective**
                - for all combinations in the Cartesian product of supported operating systems and supported python versions
                    - install `sktime` along with testing dependencies
                    - run all unit tests in `datasets` component not marked with `datadownload`
                    - upload code coverage information with relevant flags
        - `test_external_data`
            - **current name**: `test data downloaded from external sources`
            - **objective**
                - for all combinations in the Cartesian product of supported operating systems and supported python versions
                    - install `sktime` along with testing dependencies
                    - run all unit tests in `datasets` component marked with `datadownload`
                    - upload code coverage information with relevant flags
7. `.github/workflows/validate_extras.yml`
    - **current name**: `extras validation`
    - **jobs**
        - `test_installations`
            - **current name**: `extras installation`
            - **objective**
                - for all combinations in the Cartesian product of supported operating systems, supported python versions and supported component specific optional dependencies and `all_extras_pandas2`
                    - validate whether `sktime` can be installed in the selected python version on the selected operating system with selected optional dependency

## Execution

- workflows automatically triggered from pull requests targetting `main` branch
    1. `.github/workflows/test_code_quality.yml`
- workflows triggered by CRON schedule on `main` branch
    1. `.github/workflows/test_all.yml` -> `00:00 on Sunday`
    2. `.github/workflows/validate_datasets.yml` -> `00:00 on day-of-month 1`
    3. `.github/workflows/validate_extras.yml` -> `00:00 on day-of-month 1`
- workflows that can be triggered manually on any branch
    1. `.github/workflows/test_all.yml`
    2. `.github/workflows/validate_datasets.yml`
    3. `.github/workflows/validate_extras.yml`

# Runtime

1. For most runs triggered by PR's affecting a very small subset of `sktime` and only a single non-`base` component of `sktime`, new CI is expected (and observed) to be much faster than old CI.
2. As the number of changes and affected number of non-`base` components increase, number of jobs corresponding to different combinations of operating systems, python versions and optional dependencies increase over old CI. Since the environment setup itself is quite costly, old CI is observed to be faster in these cases as it uses a fixed number of jobs.
3. If there is a change in `pyproject.toml`, it introduces a significantly high number of jobs as installation of every single component-specific optional dependency (along with `all_extras_pandas2`) is validated on all supported environments (combinations of operating systems and python versions). Even though every single job itself does not take much time, this takes a significantly long time due to environment setup and limitation on number of concurrent jobs. This is a situation not covered by old CI, and can not be compared for reference.

# Known Limitations

- If a new commit is pushed in a PR while `.github/workflows/test_code_quality.yml` workflow is already running for previous commit, it triggers a new run without cancelling the previous one. It is expected to be handled using the [`concurrency` keyword](https://docs.github.com/en/actions/using-jobs/using-concurrency), but it did not work as expected when attempted in #5498. This is being tracked in #5525.
- Currently any change in `pyproject.toml` triggers a run of `.github/workflows/validate_extras.yml` workflow through `.github/workflows/test_code_quality.yml` workflow, even if it is a small text change. It can be improved with the goal that only changes in dependency or optional dependencies will trigger the workflow, and only the affected dependencies will be validated instead of all. This is being tracked in #5477.

# User Feedbacks

1. If any dependency specification is changed in `pyproject.toml`, installation of that optional dependency (and as of now, every other dependencies as well) is validated by `.github/workflows/validate_extras.yml` workflow triggerd through `.github/workflows/test_code_quality.yml` workflow. However, it does not test all estimators that are affected by that dependency, unless there is change in the corresponding modules. This is being tracked in #5292.
2. Most common CI runs are triggered from pull requests, and these mainly affect `.github/workflows/test_code_quality.yml` workflow, which also trigger other workflows as and when applicable. The names of such jobs are considerably longer and potentially be confusing for users/developers unfamiliar with the general naming format: `<name of root workflow>/<name of root workflow's job that triggers child workflow>/<name of job in child workflow> (strategy matrix combination)`. This is being tracked in #5523.

# Next Steps

@fkiraly / @benHeid / @achieveordie / @sktime/core-developers can you please review this summary and share your feedbacks for at least, but not limiting to, the following issues?

1. shorter and/or more apropriate names for workflows/jobs/actions/steps
2. retirement strategy for old CI
3. additional issues (bugs/enhancements/limitations) in new CI

Also, if it makes sense, shall we move this summary (updated with your suggestions) in `sktime` documentation?

**Comment by fkiraly** [comment]
Copying my comments from discord on the topic "shorter names":

my main concern is just to make these shorter - even if the format does not change, they still could be made shorter!
`run unit tests for individual components without extras` -> `test-comps` or similar. At least run unit tests for is completely redundant, the rest can also be shortened.

The key workflow element is "read & classify" or "read & find", so making these short and distinctive is important. Read & find is also affected by number of elements in the list.

how about `test/no-softdep/classification(3.8, ubuntu)`? 
just containing the differentiating information
I do think that's better than `validate code quality workflow/run unit tests for individual components without extras/test components without extras (3.8, windows-latest)` (which does not even contain the information that the classification set is tested) 
a list of jobs would look like:

```
test/no-softdep/classification(3.8, ubuntu)
test/softdep/classification(3.8, ubuntu)
test/no-softdep/forecasting(3.9, windows)
test/no-softdep/forecasting(3.10, windows)
```

the same list under current would look like

```
validate code quality workflow/run unit tests for individual components without extras/test components without extras (3.8, ubuntu)
validate code quality workflow/run unit tests for individual components with extras/test components without extras (3.8, ubuntu)
validate code quality workflow/run unit tests for individual components without extras/test components without extras (3.9, windows)
validate code quality workflow/run unit tests for individual components without extras/test components without extras (3.10, windows)
```

**Comment by fkiraly** [comment]
I also copy my general feedback from #5101, on useability

Sounds good, although I would like to ensure I can easily switch to new workflow (test & release) before we completely remove the old CI.

There are four issues currently, for me:

* names of jobs are too long, many jobs look the same in the GitHub CI GUI tab. I do not want to switch or complicate worfklows, so we should shorten the names.
* many jobs are spawned in some cases. Maybe this cannot be reduced, but it makes searching for single failed jobs hard
* new CI does not auto-cancel like the old CI, this creates additional clicks: https://github.com/sktime/sktime/issues/5525
* new CI always shows as failed, even if all jobs pass (in the "actions" view)

**Comment by yarnabrina** [comment]
> new CI always shows as failed, even if all jobs pass (in the "actions" view)

I was not aware of this, can you please link a PR or a specific run?

---

I'm happy to shorten the names if you please share your suggestions. I'm out of ideas with constraints of uniqueness (across workflows and jobs) and appropriateness.

**Comment by fkiraly** [comment]
> I was not aware of this, can you please link a PR or a specific run?

Checking, not all PR are examples, only some.

To reproduce:
* go to the "actions" tab of the repo
* find a CI run where old CI passes, new CI fails

example in screenshot:
![image](https://github.com/sktime/sktime/assets/7985502/753c3367-d10f-4e64-a27c-6ad979deb150)
(this one: https://github.com/sktime/sktime/actions/runs/7426396751)

**Comment by fkiraly** [comment]
> I'm happy to shorten the names if you please share your suggestions.

suggestions are here:
https://github.com/sktime/sktime/issues/5706#issuecomment-1879671293

**Comment by yarnabrina** [comment]
> 
> a list of jobs would look like:
> 
> ```
> test/no-softdep/classification(3.8, ubuntu)
> test/softdep/classification(3.8, ubuntu)
> test/no-softdep/forecasting(3.9, windows)
> test/no-softdep/forecasting(3.10, windows)
> ```
> 
> the same list under current would look like
> 
> ```
> validate code quality workflow/run unit tests for individual components without extras/test components without extras (3.8, ubuntu)
> validate code quality workflow/run unit tests for individual components with extras/test components without extras (3.8, ubuntu)
> validate code quality workflow/run unit tests for individual components without extras/test components without extras (3.9, windows)
> validate code quality workflow/run unit tests for individual components without extras/test components without extras (3.10, windows)
> ```
> 

As I tried to explain in Discord, these are not same and not corresponding to one another. The second set uses `all_extras_pandas2` to run tests in those directories for which a specific optional dependency is not created. These do not cover classification or forecasting at all, as they have their extras added in #5136.

As I said in my too long description in the original issue, no soft dependencies is never applicable for any component in current new CI. I am really not sure what else to add in the description to explain that part better. For a major component like `forecasting`, it's always tested with `sktime[forecasting,tests]`, and for a component like `distances`, it's tested with `sktime[all_extras_pandas2,tests]`. Everything (except base framework testing) needs soft dependencies, it's either specifically for that component (e.g. forecasting/clasification/etc.) or `all_extras_pandas2`. I don't know if we ought to use softdep as nosoftdep to distinguish this to help shorten names, and create confusions later.

Also, do you want to name every workflow as `test`, whether it is the workflow to test base framework, or to test components with dedicated extra (e.g. forecasting/classification/etc.), or to test all other components (e.g. pipeline/split/etc.), or to run code quality?

---

> I'm out of ideas with constraints of uniqueness (across workflows and jobs) and appropriateness.

I know I use longer and more explicit names in general, and hence definitely I think the constraints of uniqueness and appropriateness are very important. I undestand that need not be (and most defnitely it's not) applicable to others in general. If you please confirm you want these duplicating names for all jobs and workflows, I'll send a PR implementing those everywhere by tomorrow.

**Comment by fkiraly** [comment]
> As I tried to explain in Discord, these are not same and not corresponding to one another.

Apologies, but your long list does not make it clear what is exactly tested in which environment.

Just to clarify, are you saying "without extras" actually means "with all extras", i.e., installing `all_extras_pandas2`?

What would be really helpful for me if you updated each job with the exact pip install code of the environment, e.g., `sktime[dev,classification]` or similar.

**Comment by fkiraly** [comment]
> As I tried to explain in Discord, these are not same and not corresponding to one another.

I see, though I still think my point remains valid - you should try to find strings that are around length 4-8 rather than these very long names.

**Comment by fkiraly** [comment]
Moving some discussion here from closed PR https://github.com/sktime/sktime/pull/5638 to track some todos.

@fkiraly:

We will still need to resolve:
outstanding questions on how and which parts of current CI to deprecate or replace, and
https://github.com/sktime/sktime/issues/5559
https://github.com/sktime/sktime/issues/5525
https://github.com/sktime/sktime/issues/5523

Related discussion also in:
https://github.com/sktime/sktime/issues/5671
https://github.com/sktime/sktime/issues/5436
https://github.com/sktime/sktime/issues/5477
https://github.com/sktime/sktime/issues/5292

@yarnabrina:


#5559 - addressed by #5638
#5525 - not addressed yet, was attempted in #5498, but didn't work, reverted in 1dee835dcd140fe433a3dde981d9dcadbb73c3dc.
#5523 - not adressed yet, but it's more good-to-have than necessary in my opinion.

#5671 - partially addressed as `pyproject` changes would be tested, no explanation on `dependabot`'s partial attempt.
#5436 - not addressed and out-of-scope for CI, but may be that issue can be broadened to even 3.12.
#5477 - not addressed, very good-to-have to decrease runtime of new CI, open for suggestions.
#5292 - not addressed, open for suggestions, indirectly can be tested by manual runs of `test_all` workflow on that branch.

**Comment by fkiraly** [comment]
Another thing, I noticed that "validate extras", `validate_extras` or `validate-extras` seem to be redundant.

If I understand correctly - and please correct me if wrong - this just runs an install like `pip install sktime[classification]` to check that it works. As we do this anyway later, this feels duplicative, so we should remove this step.

**Comment by yarnabrina** [comment]
> Just to clarify, are you saying "without extras" actually means "with all extras", i.e., installing all_extras_pandas2?

Quoting (partially) from discord:

> This is `running tests for "components without extras"`, not `running "tests without extras" for components`.

What I mean by `components` are names of sub-directories under `sktime` directory, e.g. `base`, `datasets`, `forecasting`, etc.

What I mean by `components with extras` (or `components with dedicated optional dependencies`) are those sub-directories of `sktime` for which we have defined an "extra" in `pyproject.toml` with same name and it's sufficient to do `pip install sktime[component_name]` to use most (if not all) the functionalities defined in that sub-directory. Examples for these are `forecasting`, `classification`, etc.

What I mean by `components without extras ` (or `components that lack a dedicated optional dependency`) are those sub-directories of `sktime` for which there is no "extra" in `pyproject.toml`, and to use the functionalities one must either specifically install soft dependencies one by one manually or use the full set with `pip install sktime[all_extras_pandas2]`.

> What would be really helpful for me if, in your description, you updated each job with the exact pip install code of the environment, e.g., sktime[dev,classification] or similar.

For testing a "component with extra", e.g. `forecasting`, new CI installs `pip install sktime[forecasting,tests]`. This is done here:

https://github.com/sktime/sktime/blob/5ab219917e23233bef2ee17bce600f5996edaeec/.github/actions/test-component/action.yml#L32.

For testing "base" component, no additional soft dependencies are installed, only additional dependencies for tests are installed.

https://github.com/sktime/sktime/blob/5ab219917e23233bef2ee17bce600f5996edaeec/.github/actions/test-base/action.yml#L29

The same happens for "datasets" component.

https://github.com/sktime/sktime/blob/5ab219917e23233bef2ee17bce600f5996edaeec/.github/workflows/validate_datasets.yml#L38

https://github.com/sktime/sktime/blob/5ab219917e23233bef2ee17bce600f5996edaeec/.github/workflows/validate_datasets.yml#L77

For testing all other components, all soft dependencies (pandas v2 version) are installed along with test dependencies.

https://github.com/sktime/sktime/blob/5ab219917e23233bef2ee17bce600f5996edaeec/.github/workflows/test_components_without_extras.yml#L62

> this just runs an install like pip install sktime[classification] to check that it works.

Correct. It does just bare installation (and nothing else) for all supported environments.

> As we do this anyway later, this feels duplicative, so we should remove this step.

If a PR modifies `pyproject.toml` (what matters is modification of dependency specification, but current we don'thave a way to check that - 2nd known limitation as per original issue description), and not any other files, for examples all PR created by Dependabot, it'll pass CI without even testing the installation, and this can lead to similar situations like what happened for `pycatch22`. I chose to be on the safer side than on the convenient side.

> Moving some discussion here from closed PR https://github.com/sktime/sktime/pull/5638 to track some todos.

I thought I covered all of these in the "Known Limitations" and "User Feedbacks" sections in the original description. Did I miss anything important? If you please point out, I can edit there for better visibility.

**Comment by fkiraly** [comment]
> but current we don't have a way to check that 

we could use the runtime `git_diff` utility

**Comment by fkiraly** [comment]
for design, I think I understand better now. I think there are still some things to do:

* I feel this needs a proper design write-up in the developer docs, or it will get too complicated. You clearly have sth in mind but I keep getting repeatedly back to this and getting confused.
* I think both the writeup and the printed names should clearly state (a) which module is tested, (b) what environment/softdep set is used, (c) what tests are run.
    *  in the top post here, it is not too clear imo
    * we should also agree on common language, e.g., "code quality" should refer only to linting (you use it for entire testing workflows), or "components" should be avoided, use "module" or "dependency set" instead, whichever more accurate

From better understanding through this discussion, I think:

* we should remove "validate extras" which only installs the env. This is redundant wth later steps that do the same
* we should really make a good effort to come up with a simple write-up of the design and short names

**Comment by fkiraly** [comment]
Further question: what happens with estimators that:
* live in a module that has a special dependency set defined
* the estimator has a dependency which is *not* in the special dependency set
* but it is in `all_extras`

**Comment by yarnabrina** [comment]
> we could use the runtime git_diff utility

I wasn't aware if `git` will be able to detect if a change is made in the definition of `all_extras_pandas2` block specification etc., which is what I think is required for #5477. If it can (or did you mean some `sktime` stuff named `git_diff`? In that case, I'm not familiar with it), that'd be great to have.

> "code quality" should refer only to linting (you use it for entire testing workflows)

I do not, and I use it only for pre-commit job only. What you're referring to is the 2nd point from User Feedbacks section (tracked in #5523):

> general naming format: <name of root workflow>/<name of root workflow's job that triggers child workflow>/<name of job in child workflow> (strategy matrix combination)`

In PR's from merge request in new CI, root workflow is the workflow that runs pre commit (so named code quality), and hence you keep seeing that as the common prefix everywhere. It's firced in GHA format, and I got no response in their discussion forum when I asked how to handle this better. 

> Further question: what happens with estimators that:
> - live in a module that has a special dependency set defined
> - the estimator has a dependency which is not in the special dependency set
> - but it is in all_extras

Can you please give an example of sich?

To directly answer, this case is not covered in new CI and that estimator will not be tested (will be ignored by soft dependency check with pytest skip if logic). If such a soft dependency exists, is there any specific reason to not add it in extra-dedicated-to-component? Or, are you just talking about hypothetical cases if we fail to add everywhere consistently by mistake? 

---

Agreed on documentation, as I asked myself in the Next Steps section. However, I have no personal experience on how to write a technical documentation (definitely didn't face these much confusion ever in any office work), so don't know how to start (and I guess it takes lots of time??).

**Comment by fkiraly** [comment]
> I wasn't aware if git will be able to detect if a change is made in the definition of all_extras_pandas2 block specification etc., 

No, it cannot - but I thought the trigge condition should simply be, `pyproject.toml` was changed. These changes are rare enough to justify running all tests imo.

**Comment by fkiraly** [comment]
>> "code quality" should refer only to linting (you use it for entire testing workflows)

> I do not, and I use it only for pre-commit job only. 

I think you do, even if only by accident.

On current main, the entire new CI is called "validate code quality workflow", and the main orchestrator file for the entire new CI is called `test_code_quality`.

**Comment by fkiraly** [comment]
>> Further question: what happens with estimators that:
>> * live in a module that has a special dependency set defined
>> * the estimator has a dependency which is not in the special dependency set
>> * but it is in `all_extras`

> Can you please give an example of such?

Typically this is the case for more obscure dependencies or those with lots of secondary dependencies, I thought this was to keep the hand-curated special dependency sets smooth. Another possibility are data container related soft dependencies.

Examples:
* `arch`, this is also a dependency of `param_est`, but not tested there
* `scikit-optimize`, used in `forecasting`, but semi-abandoned and puts servere upper bounds
* `gluonts`
* `xarray`, data container

**Comment by yarnabrina** [comment]
> I think you do, even if only by accident.

> On current main, the entire new CI is called "validate code quality workflow", and the main orchestrator file for the entire new CI is called test_code_quality (the file .github/workflows/test_code_quality.yml)

I'm sorry, but I'm 100% confident that there's a communication gap. The name of the file has no role in the name you see. Since you're renaming almost everything (file name, workflow name, job name, etc.) in #5706 as "test" or similar, it may appear to be solved by file rename but that is not the case. I'm ignoring this point (and #5523) from now onwards as #5702 should meet your expectations.

---

> I thought the trigger condition should simply be, pyproject.toml was changed. These changes are rare enough to justify running all tests imo.

That's exactly what my plan was, and that's exactly what `validate_extras` does. So, I am not sure why you want to remove that workflow altogether.

---

> Typically this is the case for more obscure dependencies or those with lots of secondary dependencies, I thought this was to keep the hand-curated special dependency sets smooth.

Yes, this is a fair point, and as of now it's not covered in new CI. How do you want to address this? I can think of at least two possibilities (other than #5719 - will reply there):

1. In `test_components_without_extras`, where `all_extras_pandas2` is installed and it's triggered for all directories except specific dependency sets, we can stop ignoring. -> basically same as old CI
2. We can create `forecasting-full`, `classification-full` etc. only for the sake of testing in CI, but our recommendations will remain just `forecasting`, `classification`, etc. for end users. -> too many extras in my opinion

---

> > I was not aware of this, can you please link a PR or a specific run?
> 
> Checking, not all PR are examples, only some.
> 
> To reproduce:
> 
>     * go to the "actions" tab of the repo
> 
>     * find a CI run where old CI passes, new CI fails

I missed this earlier. I was not aware of this, and haven't seen failure in any PR pages ever (only success and skip). Is this of high priority?

**Comment by fkiraly** [comment]
> I'm sorry, but I'm 100% confident that there's a communication gap. The name of the file has no role in the name you see.

I know, but there's two similar problems:

* you named a file which contains much more than just "code quality" tests `code_quality_sth`
* in the same file, the orchestrator workflow is also called "code quality", but it also includes proper testing, also misleading

**Comment by fkiraly** [comment]
> That's exactly what my plan was, and that's exactly what `validate_extras` does. So, I am not sure why you want to remove that workflow altogether.

Oh, I see. I got confused by the naming. To confirm, the (curretly so named) `validate_extras` is triggered only if `pyproject` has changed?

**Comment by fkiraly** [comment]
>> Typically this is the case for more obscure dependencies or those with lots of secondary dependencies, I thought this was to keep the hand-curated special dependency sets smooth.

> Yes, this is a fair point, and as of now it's not covered in new CI. How do you want to address this? I can think of at least two possibilities (other than https://github.com/sktime/sktime/issues/5719 - will reply there):

Well, #5719 is for me the most programmatic solution, but possibly also the most difficult one.

> In `test_components_without_extras`, where `all_extras_pandas2` is installed and it's triggered for all directories except specific dependency sets, we can stop ignoring. -> basically same as old CI

That would reintroduce coverage, but it would double test time, no?

> We can create `forecasting-full`, `classification-full` etc. only for the sake of testing in CI, but our recommendations will remain just forecasting, classification, etc. for end users. -> too many extras in my opinion

That would also work, but indeed cause the "too many extras" problem, even with the ones we currently have dependency management is becoming inconsistent...
This would also work, though, we do not need to publicly advertise the "full" dep sets, but it sounds like a hassle still.

**Comment by fkiraly** [comment]
@yarnabrina, with #5785 merged, we should discuss next steps:

* naming - descriptivity, delineation, shortness
* removing or reducing old CI steps, optimally without reducing coverage
    * the topic of `all_extras` packages not in the curated sets - should these be considered extra? How to cover these
    * the topic of retaining coverage for `pandas 1` - e.g., move to CRON?
* coverage reports - incorrectly low coverage reports due to incremental and subsampled testing
* any dashboards for monitoring?

Anything else comes to you mind, @yarnabrina?

**Comment by fkiraly** [comment]
Also, sequencing/conditioning - to minimize overall test time if sth fails

**Comment by yarnabrina** [comment]
> * coverage reports - incorrectly low coverage reports due to incremental and subsampled testing

I just triggered a tes-all run manually. Let's see if it passes and what is the coverage.

https://github.com/sktime/sktime/actions/runs/7872869678

> any dashboards for monitoring?

What does this mean? If you mean if CI is passing or failing, I think Github provides readymade badges.

> Anything else comes to you mind

I think we should all be in agreement, or at least aligned, regarding what is the scope of CI. For example what all changes are so crucial that we should test everything without skipping (e.g. base? @benHeid suggested utils as well, may be datatypes, registry, etc.), or what all are checks we run only for compatibility purposes but we don't plan to make code changes to support issues (e.g. pandas 1 support), etc.

Another thing is what is scope of testing in general. New CI's goal is to run unit tests for specific components as independently from other components as possible. There are of course few tests that run in old CI which are more of integration type, e.g. if transformation+forecasting or network+regression is working together, etc. I said somewhere (can't find the link) that we may want to plan how can those be supported in new CI while still somehow adjusting for possible use of older versions of packages, which was one of my main complaints against old CI.

Naming is definitely another issue, you and I seem to name very differently and get confused with each others. But if we can have a file/documentation somehwere listing what is all jobs/names and what is the goal of them, I'll consider that as a lower priority than the above.

**Comment by yarnabrina** [comment]
> > * coverage reports - incorrectly low coverage reports due to incremental and subsampled testing
> 
> I just triggered a tes-all run manually. Let's see if it passes and what is the coverage.
> 
> https://github.com/sktime/sktime/actions/runs/7872869678

I just realised this is not going to help, as currently CI is unable to upload coverage reports. If you can review and merge #5930 (after #5929 if possible, though not a strict dependency), we should manually trigger test all once more and see what happens.

**Comment by yarnabrina** [comment]
FYI @sktime/core-developers: I triggered test all workflow on main after merging #5930, and now it reports an increased code coverage of ~81%.

![image](https://github.com/sktime/sktime/assets/39331844/61e51e39-d073-4081-86c3-f365d668e8e9)

Currently `main` branch coverage also matches this as it's hte last commit, but in future `complete` flag is what we should rely for best coverage out of new CI.

**Comment by fkiraly** [comment]
> Anything else comes to you mind, @yarnabrina?

Additional note, from discord discussion: "old" CI covers `dl` dependency set, new does not.
